File: apollo-runtime/src/test/java/com/apollographql/apollo/subscription/WebSocketSubscriptionTransportMessageTest.java
package com.apollographql.apollo.subscription;

import com.apollographql.apollo.response.CustomTypeAdapter;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.OperationName;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.api.ScalarType;
import com.apollographql.apollo.api.Subscription;
import com.apollographql.apollo.api.internal.UnmodifiableMapBuilder;
import com.apollographql.apollo.response.ScalarTypeAdapters;

import org.junit.Before;
import org.junit.Test;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.Map;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

import static com.google.common.truth.Truth.assertThat;

public class WebSocketSubscriptionTransportMessageTest {
  private MockWebSocketFactory webSocketFactory;
  private WebSocketSubscriptionTransport subscriptionTransport;
  private MockSubscriptionTransportCallback transportCallback;

  @Before public void setUp() throws Exception {
    webSocketFactory = new MockWebSocketFactory();
    transportCallback = new MockSubscriptionTransportCallback();

    WebSocketSubscriptionTransport.Factory factory = new WebSocketSubscriptionTransport.Factory("wss://localhost/", webSocketFactory);
    subscriptionTransport = (WebSocketSubscriptionTransport) factory.create(transportCallback);

    subscriptionTransport.connect();
    assertThat(webSocketFactory.webSocket).isNotNull();
  }

  @Test public void connectionInit() {
    subscriptionTransport.send(new OperationClientMessage.Init(Collections.<String, Object>emptyMap()));
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo("{\"type\":\"connection_init\"}");

    subscriptionTransport.send(
        new OperationClientMessage.Init(
            new UnmodifiableMapBuilder<String, Object>()
                .put("param1", true)
                .put("param2", "value")
                .build()
        )
    );
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo("{\"type\":\"connection_init\",\"payload\":{\"param1\":true,\"param2\":\"value\"}}");
  }

  @Test public void startSubscription() {
    subscriptionTransport.send(new OperationClientMessage.Start("subscriptionId", new MockSubscription(),
        new ScalarTypeAdapters(Collections.<ScalarType, CustomTypeAdapter>emptyMap())));
    assertThat(webSocketFactory.webSocket.lastSentMessage)
        .isEqualTo("{\"id\":\"subscriptionId\",\"type\":\"start\",\"payload\":{\"query\":\"subscription{commentAdded{id  name}\",\"variables\":{},\"operationName\":\"SomeSubscription\"}}");
  }

  @Test public void stopSubscription() {
    subscriptionTransport.send(new OperationClientMessage.Stop("subscriptionId"));
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo("{\"id\":\"subscriptionId\",\"type\":\"stop\"}");
  }

  @Test public void terminateSubscription() {
    subscriptionTransport.send(new OperationClientMessage.Terminate());
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo("{\"type\":\"connection_terminate\"}");
  }

  @Test public void connectionAcknowledge() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, "{\"type\":\"connection_ack\"}");
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.ConnectionAcknowledge.class);
  }

  @SuppressWarnings("unchecked")
  @Test public void data() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, "{\"type\":\"data\",\"id\":\"subscriptionId\",\"payload\":{\"data\":{\"commentAdded\":{\"__typename\":\"Comment\",\"id\":10,\"content\":\"test10\"}}}}");
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.Data.class);
    assertThat(((OperationServerMessage.Data) transportCallback.lastMessage).id).isEqualTo("subscriptionId");
    assertThat((Map<String, Object>) ((Map<String, Object>) ((OperationServerMessage.Data) transportCallback.lastMessage).payload.get("data")).get("commentAdded"))
        .containsExactlyEntriesIn(new UnmodifiableMapBuilder<String, Object>()
            .put("__typename", "Comment")
            .put("id", BigDecimal.valueOf(10))
            .put("content", "test10")
            .build()
        );
  }

  @Test public void connectionError() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, "{\"type\":\"connection_error\",\"payload\":{\"message\":\"Connection Error\"}}");
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.ConnectionError.class);
    assertThat(((OperationServerMessage.ConnectionError) transportCallback.lastMessage).payload).containsExactly("message", "Connection Error");
  }

  @Test public void error() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, "{\"type\":\"error\", \"id\":\"subscriptionId\", \"payload\":{\"message\":\"Error\"}}");
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.Error.class);
    assertThat(((OperationServerMessage.Error) transportCallback.lastMessage).id).isEqualTo("subscriptionId");
    assertThat(((OperationServerMessage.Error) transportCallback.lastMessage).payload).containsExactly("message", "Error");
  }

  @Test public void complete() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, "{\"type\":\"complete\", \"id\":\"subscriptionId\"}");
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.Complete.class);
    assertThat(((OperationServerMessage.Complete) transportCallback.lastMessage).id).isEqualTo("subscriptionId");
  }

  @Test public void unsupported() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, "{\"type\":\"unsupported\"}");
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.Unsupported.class);
    assertThat(((OperationServerMessage.Unsupported) transportCallback.lastMessage).rawMessage).isEqualTo("{\"type\":\"unsupported\"}");

    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, "{\"type\":\"unsupported");
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.Unsupported.class);
    assertThat(((OperationServerMessage.Unsupported) transportCallback.lastMessage).rawMessage).isEqualTo("{\"type\":\"unsupported");
  }

  private static final class MockWebSocketFactory implements WebSocket.Factory {
    MockWebSocket webSocket;

    @Override public WebSocket newWebSocket(@NotNull Request request, @NotNull WebSocketListener listener) {
      if (webSocket != null) {
        throw new IllegalStateException("already initialized");
      }
      return webSocket = new MockWebSocket(request, listener);
    }
  }

  private static final class MockWebSocket implements WebSocket {
    final Request request;
    final WebSocketListener listener;
    String lastSentMessage;

    MockWebSocket(Request request, WebSocketListener listener) {
      this.request = request;
      this.listener = listener;
      this.listener.onOpen(this, new okhttp3.Response.Builder()
          .request(request)
          .protocol(Protocol.HTTP_1_0)
          .code(200)
          .message("Ok")
          .build()
      );
    }

    @Override public Request request() {
      return request;
    }

    @Override public long queueSize() {
      throw new UnsupportedOperationException();
    }

    @Override public boolean send(@NotNull String text) {
      lastSentMessage = text;
      return true;
    }

    @Override public boolean send(@NotNull ByteString bytes) {
      throw new UnsupportedOperationException();
    }

    @Override public boolean close(int code, @Nullable String reason) {
      throw new UnsupportedOperationException();
    }

    @Override public void cancel() {
      throw new UnsupportedOperationException();
    }
  }

  private static final class MockSubscriptionTransportCallback implements SubscriptionTransport.Callback {
    OperationServerMessage lastMessage;

    @Override public void onConnected() {
    }

    @Override public void onFailure(Throwable t) {
    }

    @Override public void onMessage(OperationServerMessage message) {
      lastMessage = message;
    }

    @Override public void onClosed() {
    }
  }

  private static final class MockSubscription implements Subscription<Operation.Data, Operation.Data, Operation.Variables> {
    @Override public String queryDocument() {
      return "subscription{commentAdded{id\n  name\n}";
    }

    @Override public Variables variables() {
      return EMPTY_VARIABLES;
    }

    @Override public ResponseFieldMapper<Operation.Data> responseFieldMapper() {
      throw new UnsupportedOperationException();
    }

    @Override public Operation.Data wrapData(Data data) {
      return data;
    }

    @NotNull @Override public OperationName name() {
      return new OperationName() {
        @Override public String name() {
          return "SomeSubscription";
        }
      };
    }

    @NotNull @Override public String operationId() {
      return "someId";
    }
  }
}

File: apollo-runtime/src/test/java/com/apollographql/apollo/subscription/WebSocketSubscriptionTransportMessageTest.kt
package com.apollographql.apollo.subscription

import com.apollographql.apollo.api.CustomScalarAdapters
import com.google.common.truth.Truth.assertThat
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import org.junit.Before
import org.junit.Test
import java.math.BigDecimal

class WebSocketSubscriptionTransportMessageTest {
  private lateinit var webSocketFactory: MockWebSocketFactory
  private lateinit var subscriptionTransport: WebSocketSubscriptionTransport
  private lateinit var transportCallback: MockSubscriptionTransportCallback

  @Before
  fun setUp() {
    webSocketFactory = MockWebSocketFactory()
    transportCallback = MockSubscriptionTransportCallback()
    val factory = WebSocketSubscriptionTransport.Factory("wss://localhost/", webSocketFactory)
    subscriptionTransport = factory.create(transportCallback) as WebSocketSubscriptionTransport
    subscriptionTransport.connect()
    assertThat(webSocketFactory.webSocket).isNotNull()
  }

  @Test
  fun connectionInit() {
    subscriptionTransport.send(OperationClientMessage.Init(emptyMap<String, Any>()))
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo("""{"type":"connection_init"}""")
    subscriptionTransport.send(OperationClientMessage.Init(mapOf("param1" to true, "param2" to "value")))
    assertThat(webSocketFactory.webSocket.lastSentMessage)
        .isEqualTo("""{"type":"connection_init","payload":{"param1":true,"param2":"value"}}""")
  }

  @Test
  fun startSubscriptionAutoPersistSubscriptionDisabled() {
    subscriptionTransport.send(OperationClientMessage.Start(subscriptionId = "subscriptionId",
        subscription = MockSubscription(),
        customScalarAdapters = CustomScalarAdapters(emptyMap()),
        autoPersistSubscription = false,
        sendSubscriptionDocument = false))
    val expected = """{"id":"subscriptionId","type":"start","payload":{"variables":{},"operationName":"SomeSubscription","query":"subscription{commentAdded{id  name}"}}"""
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo(expected)
  }

  @Test
  fun startSubscriptionAutoPersistSubscriptionEnabledSendDocumentEnabled() {
    subscriptionTransport.send(OperationClientMessage.Start("subscriptionId", MockSubscription(),
        CustomScalarAdapters(emptyMap()), autoPersistSubscription = true, sendSubscriptionDocument = true))
    val expected = """{"id":"subscriptionId","type":"start","payload":{"variables":{},"operationName":"SomeSubscription","query":"subscription{commentAdded{id  name}","extensions":{"persistedQuery":{"version":1,"sha256Hash":"someId"}}}}"""
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo(expected)
  }

  @Test
  fun startSubscriptionAutoPersistSubscriptionEnabledSendDocumentDisabled() {
    subscriptionTransport.send(OperationClientMessage.Start("subscriptionId", MockSubscription(),
        CustomScalarAdapters(emptyMap()), autoPersistSubscription = true, sendSubscriptionDocument = false))
    val expected = """{"id":"subscriptionId","type":"start","payload":{"variables":{},"operationName":"SomeSubscription","extensions":{"persistedQuery":{"version":1,"sha256Hash":"someId"}}}}"""
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo(expected)
  }

  @Test
  fun stopSubscription() {
    subscriptionTransport.send(OperationClientMessage.Stop("subscriptionId"))
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo("""{"id":"subscriptionId","type":"stop"}""")
  }

  @Test
  fun terminateSubscription() {
    subscriptionTransport.send(OperationClientMessage.Terminate())
    assertThat(webSocketFactory.webSocket.lastSentMessage).isEqualTo("""{"type":"connection_terminate"}""")
  }

  @Test
  fun connectionAcknowledge() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, """{"type":"connection_ack"}""")
    assertThat(transportCallback.lastMessage).isInstanceOf(OperationServerMessage.ConnectionAcknowledge::class.java)
  }

  @Test
  fun data() {
    webSocketFactory.webSocket.listener.onMessage(
        webSocketFactory.webSocket,
        """{"type":"data","id":"subscriptionId","payload":{"data":{"commentAdded":{"__typename":"Comment","id":10,"content":"test10"}}}}"""
    )
    assertThat(transportCallback.lastMessage).isEqualTo(OperationServerMessage.Data(
        id = "subscriptionId",
        payload = mapOf(
            "data" to mapOf(
                "commentAdded" to mapOf(
                    "__typename" to "Comment",
                    "id" to BigDecimal.valueOf(10),
                    "content" to "test10"
                )
            )
        )
    ))
  }

  @Test
  fun connectionError() {
    webSocketFactory.webSocket.listener.onMessage(
        webSocketFactory.webSocket,
        """{"type":"connection_error","payload":{"message":"Connection Error"}}"""
    )
    assertThat(transportCallback.lastMessage)
        .isEqualTo(OperationServerMessage.ConnectionError(mapOf("message" to "Connection Error")))
  }

  @Test
  fun error() {
    webSocketFactory.webSocket.listener.onMessage(
        webSocketFactory.webSocket,
        """{"type":"error", "id":"subscriptionId", "payload":{"message":"Error"}}"""
    )
    assertThat(transportCallback.lastMessage).isEqualTo(OperationServerMessage.Error(
        id = "subscriptionId",
        payload = mapOf("message" to "Error")
    ))
  }

  @Test
  fun complete() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, """{"type":"complete", "id":"subscriptionId"}""")
    assertThat(transportCallback.lastMessage).isEqualTo(OperationServerMessage.Complete("subscriptionId"))
  }

  @Test
  fun unsupported() {
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, """{"type":"unsupported"}""")
    assertThat(transportCallback.lastMessage).isEqualTo(OperationServerMessage.Unsupported("""{"type":"unsupported"}"""))
    webSocketFactory.webSocket.listener.onMessage(webSocketFactory.webSocket, """{"type":"unsupported""")
    assertThat(transportCallback.lastMessage).isEqualTo(OperationServerMessage.Unsupported("""{"type":"unsupported"""))
  }

  private class MockWebSocketFactory : WebSocket.Factory {
    lateinit var webSocket: MockWebSocket

    override fun newWebSocket(request: Request, listener: WebSocketListener): WebSocket {
      check(!::webSocket.isInitialized) { "already initialized" }
      return MockWebSocket(request, listener).also { webSocket = it }
    }
  }

  private class MockWebSocket(val request: Request, val listener: WebSocketListener) : WebSocket {
    var lastSentMessage: String? = null

    init {
      listener.onOpen(this, Response.Builder()
          .request(request)
          .protocol(Protocol.HTTP_1_0)
          .code(200)
          .message("Ok")
          .build()
      )
    }

    override fun request(): Request = request

    override fun queueSize(): Long = throw UnsupportedOperationException()

    override fun send(text: String): Boolean {
      lastSentMessage = text
      return true
    }

    override fun send(bytes: ByteString): Boolean = throw UnsupportedOperationException()

    override fun close(code: Int, reason: String?): Boolean = throw UnsupportedOperationException()

    override fun cancel() {
      throw UnsupportedOperationException()
    }
  }

  private class MockSubscriptionTransportCallback : SubscriptionTransport.Callback {
    var lastMessage: OperationServerMessage? = null

    override fun onConnected() {}
    override fun onFailure(t: Throwable) {}
    override fun onMessage(message: OperationServerMessage) {
      lastMessage = message
    }

    override fun onClosed() {}
  }
}